### PR TITLE
KAFKA-3247: Add option to see unbalanced partitions via kafka-topic tool

### PR DESCRIPTION
**Pull request by SinghAsDev**

### CODE REVIEW

**Comment by SinghAsDev** [comment]
@gwenshap mind taking a look.

**Comment by gwenshap** [comment]
so, functionally, it looks fine. However, it is also getting to be crazy ugly. 
Any chance you can give this another thought and come up with a way to encapsulate the logic for the different options so it won't get uglier and uglier? 

Perhaps @granthenke or @ijuma have suggestions.

**Comment by SinghAsDev** [comment]
@gwenshap totally agree that it is becoming hard to read. One reason I was a bit hesitant on making a lot of changes is that TopicCommand does not have tests, so breaking things is very easy. I just filed, KAFKA-3331, to add unit tests for TopicCommand. As part of that I will make it more readable as well. Either this can go in and I will address this as part of KAFKA-3331 or I can first work on KAFKA-3331. I am fine either ways.

**Comment by gwenshap** [comment]
The only real change in this PR is to describeTopic, and the main issue is that the "if" logic is becoming hard to follow. Are you sure this cannot be improved at minimal risk without large changes to the class as a whole?

**Comment by SinghAsDev** [comment]
@gwenshap sure the if logic can be improved. However, it will still remain hard to test. I can simplify the if logic here in this patch, and we can take the testability issue to KAFKA-3331. Does that sound OK?

File: core/src/main/scala/kafka/admin/TopicCommand.scala [hunk 1]
```diff
@@ -191,12 +191,13 @@ object TopicCommand extends Logging {
     val topics = getTopics(zkUtils, opts)
     val reportUnderReplicatedPartitions = if (opts.options.has(opts.reportUnderReplicatedPartitionsOpt)) true else false
     val reportUnavailablePartitions = if (opts.options.has(opts.reportUnavailablePartitionsOpt)) true else false
+    val reportUnbalancedPartitions = if (opts.options.has(opts.reportUnbalancedPartitionsOpt)) true else false
     val reportOverriddenConfigs = if (opts.options.has(opts.topicsWithOverridesOpt)) true else false
```
Comment (ijuma) [inline core/src/main/scala/kafka/admin/TopicCommand.scala]: I haven't reviewed the PR, but can I just say that all of these `if/else` are completely unnecessary? It's enough to say `opts.options.has(opts.reportUnderReplicatedPartitionsOpt)` for example.

File: core/src/main/scala/kafka/admin/TopicCommand.scala [hunk 2]
```diff
@@ -212,9 +213,10 @@ object TopicCommand extends Logging {
             for ((partitionId, assignedReplicas) <- sortedPartitions) {
               val inSyncReplicas = zkUtils.getInSyncReplicasForPartition(topic, partitionId)
               val leader = zkUtils.getLeaderForPartition(topic, partitionId)
-              if ((!reportUnderReplicatedPartitions && !reportUnavailablePartitions) ||
+              if ((!reportUnderReplicatedPartitions && !reportUnavailablePartitions && !reportUnbalancedPartitions) ||
```
Comment (ijuma) [inline core/src/main/scala/kafka/admin/TopicCommand.scala]: Isn't this line the same as `describeConfigs` which we have already checked a few lines above?

**Comment by ijuma** [comment]
The boolean logic in this class is very weird. :) Lots of clauses and lots of negations. It should certainly be possible to make things better, but it needs a bit more time than I have to look at it right now.

File: core/src/main/scala/kafka/admin/TopicCommand.scala (hunk 1 above)
Comment (SinghAsDev) [inline core/src/main/scala/kafka/admin/TopicCommand.scala]: @ijuma I think these were added to avoid using same `opts.options.has...` at bunch of places in this class.

Comment (ijuma) [inline core/src/main/scala/kafka/admin/TopicCommand.scala]: I think you didn't understand my comment. I am not suggesting you remove the val. I am saying that you don't need `if/else` on the right-hand side:

``` scala
val reportUnderReplicatedPartitions = opts.options.has(opts.reportUnderReplicatedPartitionsOpt)
```

File: core/src/main/scala/kafka/admin/TopicCommand.scala (hunk 2 above)
Comment (SinghAsDev) [inline core/src/main/scala/kafka/admin/TopicCommand.scala]: I think this is the intention of all these long checks...

```
if just describe topic:
  for all topics:
    for all partitions:
      describeConfig
      describePartition

if describe topic with describeConfig and no (under-replicated, unavailable):
  for all topics:
    for all partitions:
      describeConfig

if describe topic for only under-replicated partitions:
  for all topics:
    for all partitions:
      if under-replicated partition:
        describePartition

if describe topic for only unavailable partitions:
  for all topics:
    for all partitions:
      if unavailable partition:
        describePartition

if describe topic for under-replicated and unavailable partitions:
  for all topics:
    for all partitions:
      if unavailable or under-replicated partition:
        describePartition
```

However, it is achieved in a complicated way.

File: core/src/main/scala/kafka/admin/TopicCommand.scala (hunk 1 above)
Comment (SinghAsDev) [inline core/src/main/scala/kafka/admin/TopicCommand.scala]: scratch that, got your point. Yea, it is not required.

**Comment by SinghAsDev** [comment]
@ijuma @gwenshap made some minor modifications to make it, hopefully, more readable. Let me know if I can improve it. Also, KAFKA-3331 should help out with simplifying and testing this.

**Comment by SinghAsDev** [comment]
@gwenshap ping for review.
